spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/SpannerTemplate.java

```diff
@@ -228,65 +228,68 @@ public long executeDmlStatement(Statement statement) {
 	}
 
 	@Override
 	public void update(Object object) {
-		applySaveMutations(this.mutationFactory.update(object, null), Collections.singletonList(object), null);
+		applySaveMutations(() -> this.mutationFactory.update(object, null), Collections.singletonList(object), null);
 	}
 
 	@Override
 	public void updateAll(Iterable objects) {
-		applySaveMutations(getMutationsForMultipleObjects(objects,
+		applySaveMutations(() -> getMutationsForMultipleObjects(objects,
 				(x) -> this.mutationFactory.update(x, null)), objects, null);
 	}
 
 	@Override
 	public void update(Object object, String... includeProperties) {
 		Set<String> incl = (includeProperties.length == 0) ? null
 				: new HashSet<>(Arrays.asList(includeProperties));
-		applySaveMutations(this.mutationFactory.update(object, incl), Collections.singletonList(object), incl);
+		applySaveMutations(() -> this.mutationFactory.update(object, incl), Collections.singletonList(object), incl);
 	}
 
 	@Override
 	public void update(Object object, Set<String> includeProperties) {
-		applySaveMutations(this.mutationFactory.update(object, includeProperties), Collections.singletonList(object),
+		applySaveMutations(() -> this.mutationFactory.update(object, includeProperties),
+				Collections.singletonList(object),
 				includeProperties);
 	}
 
 	@Override
 	public void upsert(Object object) {
-		applySaveMutations(this.mutationFactory.upsert(object, null), Collections.singletonList(object), null);
+		applySaveMutations(() -> this.mutationFactory.upsert(object, null), Collections.singletonList(object), null);
 	}
 
 	@Override
 	public void upsertAll(Iterable objects) {
-		applySaveMutations(getMutationsForMultipleObjects(objects,
+		applySaveMutations(() -> getMutationsForMultipleObjects(objects,
 				(x) -> this.mutationFactory.upsert(x, null)), objects, null);
 	}
 
 	@Override
 	public void upsert(Object object, String... includeProperties) {
 		Set<String> incl = (includeProperties.length == 0) ? null : new HashSet<>(Arrays.asList(includeProperties));
-		applySaveMutations(this.mutationFactory.upsert(object, incl), Collections.singletonList(object), incl);
+		applySaveMutations(() -> this.mutationFactory.upsert(object, incl), Collections.singletonList(object), incl);
 	}
 
 	@Override
 	public void upsert(Object object, Set<String> includeProperties) {
-		applySaveMutations(this.mutationFactory.upsert(object, includeProperties), Collections.singletonList(object),
+		applySaveMutations(() -> this.mutationFactory.upsert(object, includeProperties),
+				Collections.singletonList(object),
 				includeProperties);
 	}
 
-	private void applySaveMutations(List<Mutation> mutations, Iterable entities,
+	private void applySaveMutations(Supplier<List<Mutation>> mutationsSupplier, Iterable entities,
 			Set<String> includeProperties) {
-		maybeEmitEvent(new BeforeSaveEvent(mutations, entities, includeProperties));
+		maybeEmitEvent(new BeforeSaveEvent(entities, includeProperties));
```

Comment:
Why not include the mutations in `BeforeSaveEvent` as well? You would just need to move the next line up.

Reply:
The BeforeSaveEvent alters the entities from which the mutations are generated, so we can't pass mutations into the creation of the event.

Reply:
I see, OK.